### PR TITLE
kornia-apriltag: Make QuickDecode lookup table size configurable

### PR DESCRIPTION
**Pull request by rossng**
## 📝 Description

**Fixes/Relates to:** #450

I notice that @sidd-27 has also just opened #618 which will conflict with this PR. But I think both changes are useful and I'm happy to rebase if/once that PR is approved.

---

## 🛠️ Changes Made
- Allow customising the bit-distance to be pre-calculated for the `QuickDecode` lookup table
- Allow customising the bit-distance that is looked up at runtime
---

## 🧪 How Was This Tested?
- [x] **Unit Tests:** added new `test_quick_decode_consistent_across_hamming_configs` test to validate that different `QuickDecode` configurations are able to look up the same data. Exploited the reduced memory usage of `HammingConfig::TABLE_1_SEARCH_1` to re-enable the various `test_tag*` unit tests on aarch64.
- [ ] **Manual Verification:** I have not performed further manual verification.
- [ ] **Performance/Edge Cases:** Not yet tested. I don't expect significant runtime impact compared to master when using `HamingConfig::TABLE_2`, but other options may initialise faster and run slower (intentional!)

---

## 🕵️ AI Usage Disclosure
*Check one of the following:*
- [ ] 🟢 **No AI used.**
- [x] 🟡 **AI-assisted:** I used AI for boilerplate/refactoring but have manually reviewed and tested every line.
- [ ] 🔴 **AI-generated:** (Note: These PRs may be subject to stricter scrutiny or immediate closure if the logic is not explained).

---

## 🚦 Checklist
- [ ] I am assigned to the linked issue (required before PR submission)
- [ ] The linked issue has been approved by a maintainer
- [ ] This PR strictly implements what the linked issue describes (no scope creep)
- [ ] I have performed a **self-review** of my code (no "ghost" variables or hallucinations).
- [ ] My code follows the existing style guidelines of this project.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] (Optional) I have attached screenshots/recordings for UI changes.

---

## 💭 Additional Context
Add any other context or screenshots about the pull request here.

### CODE REVIEW

**Comment by sidd-27** [comment]
Considering that linear probing is now reasonably fast and cache-friendly, should we also provide users with the option to configure the load factor? default it to 3, but maybe as low as 2 for even more savings in space. From my rough testing, it shouldn't slow down the decode function much at all

**Comment by sidd-27** [comment]
also do benchmark the h2 search for larger families, it could be thousands of hash table lookups in the worst case

File: crates/kornia-apriltag/src/decoder.rs [hunk 1]
```diff
@@ -184,6 +286,113 @@ impl QuickDecode {
...
+    /// Looks up a codeword in the quick decode table, trying all 4 rotations.
+    ///
+    /// # Arguments
+    ///
+    /// * `rcode` - The codeword to look up.
+    /// * `entry` - Mutable reference to a `QuickDecodeEntry` to store the result.
+    pub fn lookup(&self, mut rcode: usize, entry: &mut QuickDecodeEntry) {
+        // Try all 4 rotations
+        if let ControlFlow::Break(_) = (0..4).try_for_each(|ridx| {
+            // Direct lookup (handles exact matches and pre-stored Hamming variants)
```
Comment (sidd-27) [inline crates/kornia-apriltag/src/decoder.rs]: in the most common case, ie the tag is valid with hamming 0, but simply rotated, we do unnecessary expensive lookups per rotation, maybe reorder the loops to make it so that for every rotation we first do hamming 0, then hamming 1 and then hamming 2? 

may need some benchmakring

**Comment by rossng** [comment]
> Considering that linear probing is now reasonably fast and cache-friendly, should we also provide users with the option to configure the load factor? default it to 3, but maybe as low as 2 for even more savings in space. From my rough testing, it shouldn't slow down the decode function much at all

Agree. Let's keep this PR tightly scoped - I've added #624 and can implement it.

> also do benchmark the h2 search for larger families, it could be thousands of hash table lookups in the worst case

I benchmarked the tagStandard52h13 unit test on my macbook

Config | Init (ms) | Detect (ms) | Ratio (Init/Detect)
-- | -- | -- | --
TABLE_1_SEARCH_1 | 379.29 | 72,086.61 | 0.01
TABLE_2 | 11,097.20 | 68,498.23 | 0.16
SEARCH_2 | 3.52 | 77,275.05 | 0.00

**Comment by rossng** [comment]
I've also rebased the PR.

File: crates/kornia-apriltag/src/decoder.rs (hunk 1 above)
Comment (rossng) [reply crates/kornia-apriltag/src/decoder.rs]: I suspect you're right, but I think we need a larger library of real-world test images to measure this. That's [what I started building](https://github.com/rossng/apriltag-tests) before I got sidetracked with contributing to kornia 😉

Comment (rossng) [reply crates/kornia-apriltag/src/decoder.rs]: #625

**Comment by sidd-27** [comment]
hi, check out https://github.com/kornia/kornia-rs/pull/626 . i think it should solve the Apriltag memory and initialization speed issues

**Comment by sidd-27** [comment]
@rossng  feel free to start work on the configurability of the lookup table now

**Comment by rossng** [comment]
No longer relevant due to a new approach in #626 which introduced much smaller lookup tables. It's still useful to configure the max hamming distance for detections. Tiny PR for just that part: #636
